src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```diff
@@ -0,0 +1,129 @@
+                await using var stream = await action(token);
+
+                var client = await ClientFactory.CreateAsync(options, artifactSettings, token);
+                if (stream.CanSeek) // use the stream directly
+                {                    
+                    await client.PutAsync(stream, token);
+                }
+                else // copy temporary to memory stream locally
+                {
+                    await client.UploadAsync(stream, token);
+                }
+
+                string resourceId = GetResourceId(client, options, artifactSettings);
+                return resourceId;
+            }
+            catch (AmazonS3Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        public override async Task<string> EgressAsync(
+            S3StorageEgressProviderOptions options,
+            Func<Stream, CancellationToken, Task> action,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            IS3Storage client = null;
+            string uploadId = null;
+            bool uploadDone = false;
+            try
+            {
+                client = await ClientFactory.CreateAsync(options, artifactSettings, token);
+                uploadId = await client.InitMultiPartUploadAsync(artifactSettings.Metadata, token);
+                await using var stream = new MultiPartUploadStream(client, options.BucketName, artifactSettings.Name, uploadId, options.CopyBufferSize!.Value);
+                Logger?.EgressProviderInvokeStreamAction(EgressProviderTypes.S3Storage);
+                await action(stream, token);
+                await stream.FinalizeAsync(token); // force to push the last part
+
+                // an empty file was generated
+                if (stream.Parts.Count == 0)
+                {
+                    await client.PutAsync(new MemoryStream(), token);
+                    uploadDone = true;
+                    // abort the multi-part upload
+                    await client.AbortMultipartUploadAsync(uploadId, token);
+                }
+                else
+                {
+                    await client.CompleteMultiPartUploadAsync(uploadId, stream.Parts, token);
+                    uploadDone = true;
+                }
+
+                string resourceId = GetResourceId(client, options, artifactSettings);
+                return resourceId;
+            }
+            catch (AmazonS3Exception e)
+            {
+                if (!uploadDone)
+                    await client.AbortMultipartUploadAsync(uploadId, token);
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+            catch (Exception)
```

Comment:
Please do not catch `Exception`. This looks like it was added because one of the tests throws `Exception`. Please remove this and update the test to throw `AmazonS3Exception`.